### ISSUE TITLE
[UX] Add a user-friendly parent menu browser in the content add/edit form.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#3029]
We have all been exposed to this mess. Just create a bunch of menus and menu items (using devel_generate), then add these menus as available for a content type, and finally try to either edit an existing piece of content, or create a new. Switch to the "Menu settings" vertical tab and check the "Provide a menu link" checkbox.

<img width="847" alt="screen shot 2018-04-12 at 11 34 41 am" src="https://user-images.githubusercontent.com/2423362/38651716-117ccf52-3e46-11e8-87b5-e8af72b30b7c.png">

...here's what the "Parent item" and "Weight" drop-down menus look like:

<img width="838" alt="screen shot 2018-04-12 at 11 33 41 am" src="https://user-images.githubusercontent.com/2423362/38651719-181bdfba-3e46-11e8-820b-2882578ee4e6.png">

<img width="590" alt="screen shot 2018-04-12 at 11 34 52 am" src="https://user-images.githubusercontent.com/2423362/38651723-1b61ff74-3e46-11e8-8a65-101c3b669b5e.png">

We can do better than that! Here's a couple of Drupal contributed modules that try to address this problem:

https://www.drupal.org/project/menu_select

> Replaces Drupal's stock node parent menu item selector with a hierarchal collapsible unordered list. Offers a breadcrumb preview of the menu item's location and an optional autocomplete search.

![image](https://user-images.githubusercontent.com/2423362/38652043-8d4974f4-3e47-11e8-9d81-795ec9cef1de.png)

https://www.drupal.org/project/menuux

> ...adds a menu browser to the content forms. Users don't choose "parent" and "weight" for the menu item they are adding. Instead they choose a menu, and the menu gets loaded using ajax. The users then places the menu item inside the menu by clicking on the item they want to place the menu below.

![image](https://user-images.githubusercontent.com/2423362/38652091-c7841b56-3e47-11e8-8fd6-d01ee2a2e25a.png)

https://www.drupal.org/project/menu_link_weight

> ...replaces the standard numeric weight dropdown widget for menu links in the node form with a tabledrag widget that lists all children for the selected parent.

![image](https://user-images.githubusercontent.com/2423362/38652124-e0d273f0-3e47-11e8-9b15-61679c4e1af4.png)

**Comment by klonos** [comment]
I think that the best way to solve this UX-wise would be to:

a) Replace the "Parent item" drop-down with a "Select menu" drop-down, which includes only menus; instead of the long, weird list of nested items.

b) Dynamically load a "menu item browser", which would basically be the same as the drag-enabled table in `/admin/structure/menu/manage/[menu-name]/list`. This would be updated if the user changes the selected menu from a above.

This way, the user will be able to add the menu item to a menu and select is weight relatively to the rest of the existing items within that menu, using a familiar way, and not having to deal with the weirdness that is menu weight.

**Comment by klonos** [comment]
The respective issue for drupal is https://www.drupal.org/project/drupal/issues/605894

Here's a proposed design:

![image](https://user-images.githubusercontent.com/2423362/86533938-646aeb80-bedd-11ea-8fe2-c1984fe4106f.png)

**Comment by indigoxela** [comment]
My personal favorite is https://www.drupal.org/project/menu_link_weight (making weight item a tabledrag widget).
People using Backdrop are already familiar with tabledrag, as we use it in many places. A completely new concept might be confusing.

Additionally to the node form we should also evaluate this sort of widget in the menu item form.

Any objections so far?

**Comment by klonos** [comment]
I'm also in favor of the menu_link_weight UI for the node edit form. As for the menu item edit form, I'd rather have #475 combined with something like https://www.drupal.org/project/menu_editor instead:

<img width="1059" alt="Screen Shot 2019-05-28 at 7 35 19 am" src="https://user-images.githubusercontent.com/2423362/58440194-c7575400-811b-11e9-9082-02be33f9ee12.png">

**Comment by findlabnet** [comment]
One question: where is the place for language in this hierarchy?

**Comment by indigoxela** [comment]
>  where is the place for language in this hierarchy?

That's a good point. For the node form that's no problem, because node menu items follow the node language.

But we need to take care that the menu form table doesn't get cluttered with too many columns. And I also don't like the abbreviated column headers (enabled, expanded) in the "power edit" screenshot above.
I'm sure, there's a better solution.

**Comment by klonos** [comment]
Yup, we simply cannot fit everything in a single table like that - I meant to use that screenshot as an example. I think that there would need to be only the most basic/common things exposed as columns (IMO, those should be the label, path, language, and enabled/disabled checkboxes); the rest like the description seem less important, so they would simply need to remain "hidden".

Perhaps the last column could be a "Operations" column, with a "Configure" button, which would open a dialog (similar to the way that things work with the text filters table).

Another option would be to have all the "secondary" settings for each menu item expand/collapse (same way that the more/less links work in the site status page). Menu item editing in WordPress works in a similar fashion:

<img width="1073" alt="Screen Shot 2020-07-25 at 2 12 02 pm" src="https://user-images.githubusercontent.com/2423362/88455791-16645a80-ce81-11ea-9b37-7465e618d8c2.png">

...when you expand each menu, more options are revealed:

<img width="777" alt="Screen Shot 2020-07-25 at 2 14 46 pm" src="https://user-images.githubusercontent.com/2423362/88455814-4dd30700-ce81-11ea-9c32-ace0d01a30cd.png">

It's interesting how they have this collapsible "Screen Options" panel at the top of the menu editing page, which when expanded allows defining more advanced settings for each menu item (their "Title attribute" is our menu "Description"):

![screencapture-demos2-softaculous-WordPresspvdkkjjmgl-wp-admin-nav-menus-php-2020-07-25-14_17_46](https://user-images.githubusercontent.com/2423362/88455915-29c3f580-ce82-11ea-8723-1a080a435926.png)

As you can see from the screenshots, WordPress also has this "Manage with Live Preview" option, which allows editing menus in the same fashion the color settings page works for our themes. Interesting! 🤔 ...I've tested it, and there was a tiny glitch or two, but I did appreciate the ability to view my changes live. I imagine that it's very useful for menus with a few items, but it is not ideal when wanting to mass-edit menus with a large amount of menu items.

**Comment by olafgrabienski** [comment]
Referring to https://github.com/backdrop/backdrop-issues/issues/4466#issuecomment-662474477 and trying to avoid pitfalls, I'm mentioning a common use case on my sites: longer sub-menus where the items are to be listed alphabetically by default, which works because the default menu item weight of `0`. So, please don't change this default without considering my use case.

**Comment by indigoxela** [comment]
Out of curiosity I ported menu_link_weight module to Backdrop (locally, not published on github).
We could as well keep this functionality in contrib, if the risk to break something for users would be too high.

**Comment by indigoxela** [comment]
I guess, the first thing to figure out here is: core or contrib.

Personally, I'm leaning towards porting menu_link_weight and keeping it in contrib.
We can then observe the amount of installs and if it goes beyond a certain amount, incorporate it in core.

Background to that: while many of us think that the current handling is awful, this might not be what most users think. I'm aware that Backdrop (rightly!) decides things to improve UX in general. I'm just not convinced that this is really the case here.
I'd rather suggest to let the community decide (by the number of installs of the contrib module).

**Comment by olafgrabienski** [comment]
> Personally, I'm leaning towards porting menu_link_weight and keeping it in contrib at first.

I agree because I think it's not easy to make this really user-friendly. Curious about the usability of *Menu Link Weight* and how it works in different situations, e.g. with many menu items.

Also, if we put or implement it in core later, we may already have identified problematic aspects.

**Comment by indigoxela** [comment]
Here's the contrib repo: https://github.com/backdrop-contrib/menu_link_weight

I admit, that might have been a bit premature, but it's always easier to decide based on something more concrete.

**Comment by klonos** [comment]
> We can then observe the amount of installs and if it goes beyond a certain percentage, incorporate it in core.

I generally agree that usage metrics is a very nice way to present tangible evidence that a certain feature is what the majority likes. In fact, very often you'll see me refer to d.org usage stats of modules, each time we are to either port something or merge it into core. Having said that though, I would like to make a couple of points:

1. We often make the assumption that if a module improves the UI/UX, it will "organically" become popular. That may be true in some cases, but we're assuming that people will figure out that such a module exists to begin with; and we then have to wait for usage numbers to go up, which may take years. In the meantime, the UX of the admin UI suffers, and in this case it's been suffering since the release of Drupal 7 (which was 9 years ago), or even earlier.

2. Some very useful contrib "utility" modules never reach a considerable usage count in the entire lifetime of a Drupal version. I'm sure there are other examples, but case in point:
   - https://backdropcms.org/project/usage/options_element (roughly any usage ever recorded in Backdrop) https://www.drupal.org/project/usage/options_element (peaked at 35k installations, which is roughly ~5% of D7 sites).
   - The 7.x versions of https://www.drupal.org/project/usage/linkit peaked at ~50k usage, which is ~7% of D7 sites.
   If we solely relied on usage stats for those, we'd never have #3764 and #1005 in core 🤷 .

As you said @indigoxela: "Backdrop (rightly!) decides things to improve UX in general." 😉

**Comment by klonos** [comment]
Thanks for starting the port in contrib @indigoxela 🎉 ...can you please consider releasing an alpha or beta, so we can easily test this with https://backdropcms.org/demo?

**Comment by indigoxela** [comment]
> can you please consider releasing an alpha or beta

- I don't know how to do that (does the packager script support that?).
- As we're discussing here "core vs. contrib" it might not smart to do a release in contrib.

Downloading a zip file is, of course, possible: https://github.com/backdrop-contrib/l10n_client/archive/1.x-1.x.zip

**Comment by klonos** [comment]
> Downloading a zip file is, of course, possible

Sure 👍 

>I don't know how to do that (does the packager script support that?).

You simply tag the release with 0 as it major number, like 1.x-0.0.1. Then when you create a release on GitHub, you can mark it as a pre-release:

![Screen Shot 2020-07-30 at 4 44 36 pm](https://user-images.githubusercontent.com/2423362/88930453-3088a900-d284-11ea-8ab8-d60dfc1ce7b7.png)

**Comment by indigoxela** [comment]
@klonos can you please remove the unrelated github screenshots? They're distracting here.

**Comment by olafgrabienski** [comment]
> https://github.com/backdrop/backdrop-issues/issues/3029#issuecomment-665659850: common use case on my sites: longer sub-menus where the items are to be listed alphabetically by default, which works because the default menu item weight of `0`

I've just tested [Menu Link Weight](https://github.com/backdrop-contrib/menu_link_weight). With the module, the handling of menu link order in the node form is super easy, really like it!

Unfortunately, once you open a node form and go the the *Menu settings* tab, the module sets its own weight values, i.e. `-50`, `-49`, `-48`, ecc. This is even the case if the items had a weight of `0`, e.g. set via the *Edit links* page of a menu. So, my use case described above wouldn't work. It would be nice to adapt the module's behavior respectively, if we implement it in core.

**Comment by indigoxela** [comment]
> Unfortunately, once you open a node form and go the the Menu settings tab, the module sets its own weight values ... So, my use case described above wouldn't work.

Yep. that's one of the reasons, I'd recommend to keep that in contrib.

**Comment by olafgrabienski** [comment]
> that's one of the reasons, I'd recommend to keep that in contrib.

Sounds reasonable! And if we consider to integrate it in core one day, we can always try to adjust the behavior (keep weight default of `0`, if that works good usability-wise), or make the functionality optional per menu.

**Comment by jenlampton** [comment]
Just getting up to date here, Has this issue become menu_link_weight in core? If so I am completely in favor! I love that module and believe will be a huge editor experience improvement :)

If we get https://github.com/backdrop/backdrop-issues/issues/4466 in first, would that cause any issues with this issue?
